### ISSUE TITLE
Architecture: final class must implement an interface (mocking, DI-ing)

### ISSUE THREAD

**Issue by kalessil** [kalessil/phpinspectionsea#341]
https://twitter.com/kalessil/status/870691669204586496

**Comment by Fleshgrinder** [comment]
Absolutely not, you forget value objects and entities. There is almost never a reason for them to have an interface, and mocking is especially not an argument, because [mocking is a code smell](https://duckduckgo.com/?q=mocking+is+a+code+smell) (see also [wewut](https://leanpub.com/wewut)).

**Comment by kalessil** [comment]
Value objects: agree, already mentioned in Twitter (entities are).

Mocking: probably, but for services (db/rest/soap interaction) I'd prefer mocking over stub-classes. 
But if final class has no public methods (ignoring setters/getters) - then I think your point will be covered.

This should be pretty usable, what do you think @Fleshgrinder

**Comment by Fleshgrinder** [comment]
Infrastructure (DB/REST/SOAP) should always come along with a ready made fake/null implementation by the vendor. Anything else is not reliable over time. Mocking has its uses, but they are very, very, very seldom (e.g. spying).

The term service is unclear to me, it is an overloaded term with too many meanings in different architectures.

A value object with a setter would also be broken by design in my book. 😉 However, a value object without a single public method also seems pretty useless to me, and sounds more like DTO rather than value object. Every object in our domain should have behavior, otherwise it’s questionable whether it should exist in the first place.

In other words: Infrastructure should **always** have an interface, but it is probably impossible to write a sniff for that. With everything else _it depends._

**Comment by kalessil** [comment]
Infrastructure: should, and the inspection should spot places where this requirement (interface, testing components) is not full-filled (we have to mock -> we should spot the mocking as root cause). 

Basically, we need only to identify value objects and DTOs - and the rest will be a broad set of services, facades, decorators, factories & co which normally are using interfaces.

Since we know how to identify DTOs and we are sleeping into DDD (or?), I wondering if final used to define value object or not, is it? 

PS: NPEs are released, you wanted to give a reference to our project )

**Comment by Fleshgrinder** [comment]
Value object are not defined by being `final`. They often are, but it is not a requirement. Such an inspection will lead to interface explosion and deep inheritance chains. It’s hard for most developers to understand the proper usage of interfaces in the first place, this will just make the situation worse.

The terms value object, entity, repository, etc. are originating from DDD, yes. However, their concepts are older and well known, Eric Evans just gave them memorable names.

> PS: NPEs are released, you wanted to give a reference to our project )

I tried it, but it does not seem to trigger:

```php
<?php

/** @return null|string */
function f(): ?string {}

$v = f()[0];
```

**Comment by kalessil** [comment]
I was researching the criteria last 2 days, but have to agree - there are no reliable criteria. The inspection can be applied only to final classes used as function/methods parameter types.

The NPEs inspection applied to nullable objects only, not scalars:
- [ ] method/field reference;
- [ ] dispatching into function/method;
- [ ] array access (`<whatever>[<whatever>]`) pattern support needs to e added;

Regarding nullable scalars, do you think it worth supporting it - I guess 99% of NPEs are related to objects only.

**Comment by kalessil** [comment]
After research we are getting back to #265

**Comment by Fleshgrinder** [comment]
#265 is much better, yes. 👍 

I think that it makes sense to implement the nullable check for any and every situation. It is a valid pattern after all, and PHP will most probably never be able to inspect that stuff (simply to heavy at runtime).
